### ISSUE TITLE
PowerShell 7 preview 5 takes way too long to launch, far too slow

### ISSUE THREAD

**Issue by JasonFossen** [PowerShell/PowerShell#10983]
PROBLEM:
When opening PowerShell 7 Preview 5, it takes 6-10 seconds before the prompt/cursor appears so that commands may be entered.  This is embarrassingly slow.  If it is not some quirk of my VM, this must be fixed before GA.

REPRODUCE:
Fresh install of Windows Server 2019 Datacenter, version 1809, OS build 17763.737.
Network adapter enabled inside the VM.
VM is on an isolated virtual segment, no outside LAN access.
This Windows Server VM is its own DNS server.
Set POWERSHELL_TELEMETRY_OPTOUT = 1 as a system environment variable.
Install PowerShell Core 7.0.0-preview.5 x64.
Reboot and log on as Administrator.
Run App:  pwsh.exe -noprofile
Must wait 6-9 seconds.
Confirm that POWERSHELL_TELEMETRY_OPTOUT = 1 is correctly set in pwsh.
Disable the VM network adapter inside the VM.
Run App: pwsh.exe -noprofile
Launches normally, less than 1 second delay.

NOTES:
PowerShell appears to be doing DNS lookups for ctldl.windowsupdate.com and api.github.com, possibly for telemetry spying, even though POWERSHELL_TELEMETRY_OPTOUT has been set to 1 in the Control Panel System app.  The shell is blocking until the DNS lookups fail.

The telemetry spying is bad enough by itself, but the 8-second delay makes it unshippable, especially when telemetry opt out is configured.  Many networks, like for the DoD, must run air-gapped.

**Comment by SteveL-MSFT** [comment]
What you are seeing is calls for update notification as detailed in this RFC: https://github.com/PowerShell/PowerShell-RFC/pull/162/files which includes instructions on how to disable it.  Please don't state or imply any "spying" is going on.  We deliberately publish RFCs being completely transparent on all data that is sent from the product.  Since PowerShell is Open Source, you can see exactly what is being sent and where.

Also, the first cold start of PowerShell, it spends time building the module analysis cache so it knows what commands and cmdlets are available.  This is only on the first cold start.

**Comment by JasonFossen** [comment]
Thanks for the quick response.

However, setting POWERSHELL_UPDATECHECK to "Off" as a system environment variable does not decrease the startup time of pwsh, it's still about 8 seconds.  Disabling the VM's network interface reduces launch time to less than 1 second.  Is it a different setting?  Am I reading the RFC correctly? 

(Regarding telemetry on non-beta products: if it's opt-in, it's feedback; if it's out-out, it's spying.  But this isn't the place for another debate on the issue, I'm only interested here in the startup delay.)

**Comment by SteveL-MSFT** [comment]
Looks like the code currently expects `POWERSHELL_UPDATECHECK_OPTOUT=1` to disable.  However, that check is supposed to run on a background thread not impact startup perf.  cc @daxian-dbw

**Comment by xsoheilalizadeh** [comment]
Also it getting worse when you load some modules such as posh-git in the profile.

**Comment by daxian-dbw** [comment]
@JasonFossen The RFC was updated after preview.5 to make the update check behavior more than just on and off, but the code in preview.5 was using `POWERSHELL_UPDATECHECK_OPTOUT` similar as the telemetry, so you can set `POWERSHELL_UPDATECHECK_OPTOUT` to 1 to disable the update notification functionality.

But at any rate, the network query to `GitHub.com` happens on a background thread and shouldn't affect the startup time. I will setup the environment as you described and try to repro the issue.

**Comment by JasonFossen** [comment]
@daxian-dbw  Thank you.  Btw, setting POWERSHELL_UPDATECHECK_OPTOUT to 1 as a system environment variable makes no difference on my VM.  However, disabling recursion on my DNS service in my VM (DNS properties > Advanced tab) also reduces the pwsh startup delay to less than 1 second.  My VM is a DNS client to itself.  Disabling the VM's network interface has the same effect.

**Comment by iSazonov** [comment]
@JasonFossen Please share follow information after you run pwsh:
```powershell
 [environment]::GetEnvironmentVariables()
```

```powershell
[environment]::GetEnvironmentVariable("POWERSHELL_UPDATECHECK_OPTOUT")
```

**Comment by JasonFossen** [comment]
@iSazonov 

Hello Ilya, these are the names being queried most consistently after launching "pwsh.exe -noprofile":

- ctldl.windowsupdate.com
- checkappexec.microsoft.com
- az764295.vo.msecnd.net

However, when I disable recursion support on the DNS server, only the first name is queried (ctldl.windowsupdate.com), then, when the DNS server refuses to resolve it, pwsh immediately displays a working cursor/prompt, no delay.  When recursion support is enabled on the DNS server again, ctldl.windowsupdate.com is queried repeatedly until lookup failure, and then pwsh displays a working cursor/prompt.  The other two names are not as consistently queried.  Earlier, I saw a query for api.github.com too, but I don't see that anymore.  

The DNS server is the local VM on which pwsh is running.  The VM is air-gapped from the physical LAN.  Queries were captured with Wireshark on the VM.  The VM is running Windows Server 2019 Datacenter Evaluation v1809, OS build 17763.737, installed on 3-Nov.2019 as a VM in VirtualBox 6.0.14.


 `[environment]::GetEnvironmentVariables() | Select *`

NUMBER_OF_PROCESSORS            : 4
HOMEPATH                        : \Users\Administrator
DriverData                      : C:\Windows\System32\Drivers\DriverData
FPS_BROWSER_APP_PROFILE_STRING  : Internet Explorer
USERNAME                        : Administrator
windir                          : C:\Windows
CommonProgramFiles(x86)         : C:\Program Files (x86)\Common Files
PROCESSOR_REVISION              : 3c03
PROCESSOR_IDENTIFIER            : Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
USERDNSDOMAIN                   : TESTING.LOCAL
**POWERSHELL_UPDATECHECK_OPTOUT   : 1**
FPS_BROWSER_USER_PROFILE_STRING : Default
PUBLIC                          : C:\Users\Public
POWERSHELL_DISTRIBUTION_CHANNEL : MSI:Windows Server 2019 Datacenter Evaluation
PATHEXT                         : .COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL
ProgramFiles                    : C:\Program Files
PROCESSOR_ARCHITECTURE          : AMD64
SystemRoot                      : C:\Windows
COMPUTERNAME                    : WIN-F8ANUMHVJH5
LOGONSERVER                     : \\WIN-F8ANUMHVJH5
HOMEDRIVE                       : C:
TMP                             : C:\Users\ADMINI~1\AppData\Local\Temp
ProgramFiles(x86)               : C:\Program Files (x86)
APPDATA                         : C:\Users\Administrator\AppData\Roaming
PROCESSOR_LEVEL                 : 6
CommonProgramW6432              : C:\Program Files\Common Files
ProgramData                     : C:\ProgramData
LOCALAPPDATA                    : C:\Users\Administrator\AppData\Local
SystemDrive                     : C:
PsNotificationPreference        : Off
ProgramW6432                    : C:\Program Files
**POWERSHELL_TELEMETRY_OPTOUT     : 1**
USERDOMAIN_ROAMINGPROFILE       : TESTING
CommonProgramFiles              : C:\Program Files\Common Files
OS                              : Windows_NT
TEMP                            : C:\Users\ADMINI1\AppData\Local\Temp
PSModulePath                    : C:\Users\Administrator\Documents\PowerShell\Modules;C:\Program Files\PowerShell\Modules;c:\program files\powershell\7-preview\Modules;C:\Windows\system32\WindowsPowerShell\v1.0\Modules
USERPROFILE                     : C:\Users\Administrator
ALLUSERSPROFILE                 : C:\ProgramData
Path                            : C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\PowerShell\7-preview\preview;C:\Program Files\Microsoft VS Code\bin;C:\Users\Administrator\AppData\Local\Microsoft\WindowsApps
ComSpec                         : C:\Windows\system32\cmd.exe
USERDOMAIN                      : TESTING

**Comment by iSazonov** [comment]
> ctldl.windowsupdate.com

This site is used for updating root certificates.

I guess there are three issue:
1. DNS. If you enable recursion your DNS service does all work itself. Perhaps you need to update root DNS site list. Or you get a broken response from a stolen DNS.
2. You could look logs (CAPI) to discover why your server has not an updated root certificate list.
3.  I don't know why pwsh causes the request every time. I guess it could be third-party AV soft. You could remove the AV and look a result.

**Comment by SteveL-MSFT** [comment]
pwsh calls into amsi API (anti-malware) to validate scripts which calls into the AV software that's installed.  This is probably triggering the attempt to see if there's new certs for @iSazonov's 3rd point.

**Comment by JasonFossen** [comment]
Note that, in the VM, the only AV is the built-in Windows Defender and its realtime scanning had been disabled with:

`Set-MpPreference -DisableRealtimeMonitoring $True -Force`

PowerShell Core is being launched with "pwsh.exe -noprofile" and no third-party modules are installed.  

DNS is the symptom, not the cause, of the startup delay.  Like Dongbo mentioned, it's probably a thread in pwsh launch/init that is supposed to be running in the background, but is blocking instead.  I assume this is what Dongbo is looking into now.

**Comment by daxian-dbw** [comment]
What I was looking into was whether the updates notification query to `GitHub.com`, which happens on a background thread, could be the cause.
However, given that the long delay remains after you opt'ed out the update notification (`pwsh` won't query for `GitHub.com` at all in this case), the notification update feature is likely not the cause of the symptom.

I did a bit research, `ctldl.windowsupdate.com` is for updating certificate revocation lists; `checkappexec.microsoft.com` [has something to do with Windows Defender SmartScreen](https://answers.microsoft.com/en-us/windows/forum/windows_10-security/any-win10-smartscreencanonicalizerucsuritcs-link/dee54731-4e8c-4819-8f2f-3e5696b29ff7); `az764295.vo.msecnd.net` is also for [detecting malicious code](https://github.com/Microsoft/vscode/issues/70227#issuecomment-473839377).

Like @SteveL-MSFT mentioned above, the AMSI APIs are hooked up with Windows Defender by default, and could be the trigger of those queries. /cc @TravisEz13 for some insights.

@JasonFossen Can you please try using the `7.0.0-preview.4` version of `pwsh` on the same VM and see if it has the same problem?

**Comment by JasonFossen** [comment]
@daxian-dbw  Unfortunately, no luck, 7.0.0-preview.4 also shows the same startup delay.  

Note that I didn't uninstall preview5 and then install preview4, I started with a fresh VM snapshot, no pwsh at all, then installed preview4. 

Btw, in case it makes a difference, the way I'm installing pwsh is from within ISE as local admin:

`msiexec.exe /i C:\Full\Path\PowerShell-7.0.0-preview.4-win-x64.msi /qn`

And the Windows Server installation ISO was downloaded on 3.Nov.2019:  17763.737.190906-2324.rs5_release_svc_refresh_SERVER_EVAL_x64FRE_en-us_1.iso

No patches or updates have been applied.  No third-party software installed.  The VM has never had outside network access.  If I uninstall the DNS role, or if the VM is configured with no DNS servers at all as a client, then there's no pwsh startup delay because all DNS queries are failing immediately.

I also set Windows Defender back to the default realtime monitoring with the following, but it didn't help:

`Set-MpPreference -DisableRealtimeMonitoring $False -Force`

**Comment by TravisEz13** [comment]
I don't think that disables smartscreen.  You would have to do this:
https://www.maketecheasier.com/disable-smartscreen-filter-windows10/

**Comment by JasonFossen** [comment]
@TravisEz13  Thanks Travis, I used all three methods described in that article, at least as far as Windows Server 2019 goes, rebooted, but it didn't help:  pwsh-preview4 still opens with an 8-second delay as it waits on recursive DNS queries.  Presumably preview5 would do the same.  Thank you though.

**Comment by TravisEz13** [comment]
How about this?  
https://blogs.msdn.microsoft.com/winsdk/2015/02/10/reducing-the-time-to-perform-a-crl-check-on-isolated-networks/

**Comment by JasonFossen** [comment]
Thanks @TravisEz13, it might help, I didn't try it, but the fundamental problem is not that CRL checking is slow, the problem is that the pwsh process launch likely has an init/launch/createrunspace/etc thread which should be running parallel in the background to the other threads, but instead is blocking the whole shell.  Launching a new pwsh shell should present a cursor to the user as quickly as possible, even if pwsh is doing other housekeeping in the background.  Performing _any_ kind of DNS query for any purpose whatsoever should not be in the "critical path", so to speak, of launching a new "pwsh.exe -noprofile" shell with only generic default modules installed.

**Comment by TravisEz13** [comment]
@JasonFossen Please provide a stack trace showing what you are saying.  Base on the information you have provided so far, I am assume that it is critical path security code.

**Comment by TravisEz13** [comment]
This article talk about the specific problem.  It can happen when we validate any local certificate, and is a WINDOWS function:
https://support.microsoft.com/en-us/help/2677070/an-automatic-updater-of-untrusted-certificates-is-available-for-window

**Comment by JasonFossen** [comment]
Thanks @TravisEz13 for the links, but you are not seeing the pwsh startup delay as described above?  This is on a fresh, air-gapped Windows Server VM, no updates applied, NIC enabled in the isolated VM, DNS role installed in the VM, the VM is a DNS client to itself in the NIC properties, and a clean install of pwsh7-preview5.  No stack trace is needed to see the pwsh startup delay of 8 seconds.  As the old joke goes, if the problem can't be reproduced, it's not really a problem...  ;-)

**Comment by JasonFossen** [comment]
Interesting, if the OpenSSH service is installed and pwsh7-preview4 is configured as an OpenSSH "subsystem" in the sshd_config file, then there is no startup or connection delay when connecting with:

`Enter-PSSession -HostName localhost -SSHTransport -UserName administrator`

When pwsh is run by the sshd.exe service, is there something different about running "pwsh.exe -sshs" by the service that avoids the launch delay?  Different launch code path?  System SID?

**Comment by TravisEz13** [comment]
My guess is that the call that cause the delay is deferred in the SSH scenario.

**Comment by daxian-dbw** [comment]
It looks to me the certificate revocation list (CRL) request is triggered by loading `PSReadLine` module. `PSReadLine.psm1` contains a signature block and the content of the file is scanned by the AMSI API.
In the ssh ssh server mode, `PSReadLine` will not be loaded at all, so you don't see the CRL request.

**Comment by daxian-dbw** [comment]
We possibly can move the `PSConsoleHostReadLine` function from `PSReadLine` to a C# cmdlet to avoid the delay in the `pwsh -noprofile` scenario, but it doesn't solve the problem. As soon as you load another script module with a signature in it, you will see the delay again.

**Comment by SteveL-MSFT** [comment]
PSReadLine is already catalog signed, perhaps we should individually signing the psm1 file.

**Comment by daxian-dbw** [comment]
> we should individually signing the psm1 file.

@SteveL-MSFT do you mean **skip** individually signing the psm1 file?

**Comment by SteveL-MSFT** [comment]
@daxian-dbw yeah, a word was missing there :)

If other modules the user installs have sig blocks, then that's not something we can solve.  We can remediate it for modules we own.

**Comment by iSazonov** [comment]
We could look on the issue from another side. Why do we check a sig? If PowerShell has not white certificate list (and it do has not) PowerShell can not trust to the sig, it only can check certificate revocation and then code integrity. I mean that checking a sig makes sense only at install time. To check the trust Windows has other subsystems like SRP/AppLocker.

**Comment by TravisEz13** [comment]
@SteveL-MSFT Catalog signing is not a valid for of authenticode signing for non-windows components.  Meaning the module would not work in allSigned

**Comment by TravisEz13** [comment]
@iSazonov It's not PowerShell checking the signature.  As I've said many times, it is windows.

**Comment by rkeithhill** [comment]
> Meaning the module would not work in allSigned.

Well, obviously that is not acceptable. :-) The bummer here is that PSRL is 99.9% a **binary module**.  The only script requiring signing is this one function:
```
function PSConsoleHostReadLine
{
    Microsoft.PowerShell.Core\Set-StrictMode -Off
    [Microsoft.PowerShell.PSConsoleReadLine]::ReadLine($host.Runspace, $ExecutionContext)
}
```
As @daxian-dbw says "can move the PSConsoleHostReadLine function from PSReadLine to a C# cmdlet to avoid the delay in the pwsh -noprofile scenario"`.  While it doesn't solve the issue for other modules, I don't see any other modules in a new session that require sig-checking other than PSRL. So in a way, this **would** solve the PowerShell startup time problem.  That said, I'm not sure how/when catalog sigs are checked.

@JasonFossen what startup perf do you get with `pwsh -nop -noni`? That will avoid loading PSRL.

**Comment by iSazonov** [comment]
> If I uninstall the DNS role, or if the VM is configured with no DNS servers at all as a client, then there's no pwsh startup delay because all DNS queries are failing immediately.

What if you assign Google DNS 4.4.4.4 on the network interface?

**Comment by JasonFossen** [comment]
Hi @rkeithhill, thanks for the useful tip, there is no startup delay when running 'pwsh -noprofile -noninteractive.'  However, what I'm actually concerned about in posting this issue to GitHub is PowerShell Core's market share, not my own testing VMs.  

As the 'good word' spreads about pwsh7's compatibility and performance, more people will try it out for the first time, especially in comparison to prior versions of pwsh.  Thousands of these people, however, will be in air gaps (like half the DoD), or using isolated testing VMs, or sitting at training computers, or otherwise installing pwsh7 on systems that are clients to DNS servers which do not respond correctly or fast enough.  These people are not going to blame DNS, AMSI, Windows Defender, Authenticode or CRLs for pwsh's slow startup speed.  Every time they launch pwsh and have to wait 8 seconds, they are going to think, "PowerShell Core sucks."  Then some of them will tell their friends and post on social media how "PowerShell Core sucks."  And for Linux admins in particular, who are accustomed to terminals opening more-or-less instantaneously, an 8-second delay is an _eternity_.  When they figure out that it's because pwsh is querying DNS to "phone home" back to Microsoft (the reason why won't matter), it's not going to improve their opinion of PowerShell or Microsoft, to put it mildly.  So, overall, I know we're talking about a small percentage of potential pwsh7 users, but why have this unnecessary, self-imposed issue at all?

**Comment by rkeithhill** [comment]
Could you try this right after you've started with `pwsh -nop -noni`:
```
measure-command { ipmo Microsoft.PowerShell.Management }
```
How long does that take?  Then restart with `pwsh -nop -noni` and run:
```
measure-command { ipmo PSReadLine }
```
You mention this being an issue on Linux.  I don't think any of the Authenticode cert checking applies to Linux.  In fact, execution policy (AllSigned, RemoteSigned, etc) does not apply at all on non-Windows OSs AFAICT.

**Comment by JasonFossen** [comment]
`measure-command { ipmo Microsoft.PowerShell.Management }`

9.95 milliseconds avg in 3 new shells 

`measure-command { ipmo PSReadLine }`

9.52 seconds avg in 3 new shells (recursion enabled on DNS server)
83.1 milliseconds avg in 3 new shells (recursion disabled on DNS server) 

Good point about Linux, I've only been testing pwsh7 on Windows so far, I'll try it out!

**Comment by TravisEz13** [comment]
@rkeithhill The catalog signatures on modules are checked when you upload **to** the gallery.
We really need to move to nuget package signing which works cross platform.

**Comment by rkeithhill** [comment]
@JasonFossen One last test if you don't mind.  Go into `C:\Program Files\PowerShell\7-preview\Modules\PSReadLine` and make a backup copy of `PSReadLine.psm1`.  Then open the original `PSReadLine.psm1` and rip out the signature block from beginning to end and save that file.  Then run `pwsh -nop -noni` followed by `measure-command { ipmo PSReadLine }`.  In theory, that should be fast.  Note that this would fail if your execution policy were set to `AllSigned`.

**Comment by JasonFossen** [comment]
@rkeithhill  It was fun to try, but no luck:  9.39 seconds avg, 3 tests, execution policy = bypass.   :-\

**Comment by rkeithhill** [comment]
Thanks. Then perhaps it's not the module signature but the Authenticode signature on `Microsoft.PowerShell.PSReadLine2.dll`.  That kind of makes sense because the Management & Utility modules (their psd1 files), which load fast, are also module signed.  Not sure how to verify that though. 
 
We had a similar issue on a licensing project I worked on 12 years ago. It was checking the calling assembly's Authenticode signature and on air-gapped systems (never connected to the internet - common for some of our DoD subcontractor customers), there would be this ~30 delay in that code.  I don't remember exactly how we addressed that.

**Comment by JasonFossen** [comment]
@rkeithhill  Fortunately there is no startup delay, as you suggested, for pwsh7-preview5 on Ubuntu 18.04 on the same isolated network using the same DNS server that has recursion enabled.  There were DNS queries for api.github.com and dc.services.visualstudio.com, but these were in the background and did not block the shell. No OPTOUT environment variables were set and PSReadLine loaded normally.

**Comment by SteveL-MSFT** [comment]
It seems that we should minimally provide some information to the user if there is a delay on startup with an aka.ms link to documentation explaining what is happening and what options they have

**Comment by daxian-dbw** [comment]
> That kind of makes sense because the Management & Utility modules (their psd1 files), which load fast, are also module signed.

@rkeithhill There is an optimization to the psd1 files. If the psd1 file contains a pure Hashtable expression and it's considered safe (by using the `IsSafeValueVisitor`), then the security check is skipped for the psd1 file. The psd1 files for Management & Utility modules are skipped because of this. But you are right, from what @JasonFossen described, loading `Utility` module still triggers the query, and that suggests the dll signature is playing a role here.

**Comment by TravisEz13** [comment]
@JasonFossen I profiled this and verified this is a security check of code we are actually running and if you would like to speed up this check, please adjust the settings as described here:
https://github.com/PowerShell/PowerShell/issues/10983#issuecomment-551976417

**Comment by JasonFossen** [comment]
@TravisEz13  Thank you Travis, that is useful.  Unfortunately, it will only be useful to new pwsh users who research the issue (maybe finding it here).  I think most people who encounter the problem will simply blame PowerShell Core and go back to Windows PowerShell, which has no startup delay.

**Comment by TravisEz13** [comment]
@JasonFossen We are considering mitigations.  Currently, the best fix, without changes to Windows or removing features is to notify uses of the situation.

**Comment by TravisEz13** [comment]
Here is the profile:
![image](https://user-images.githubusercontent.com/10873629/68814679-42f77f80-062e-11ea-9ce5-949736f8deb5.png)

**Comment by iSazonov** [comment]
For reference #10309

**Comment by TravisEz13** [comment]
https://github.com/PowerShell/PowerShell/issues/10983#issuecomment-553759588 this wasn't CDXML related.

**Comment by TravisEz13** [comment]
@JamesWTruher The changes to fix this would be significant.  Some possible mitigations would require an RFC.  I've opened as discussion issue it to get feedback:  https://github.com/PowerShell/PowerShell/issues/11074
I'm going to close this issue as the discussion should track this going forward.

**Comment by iSazonov** [comment]
> [#10983 (comment)](https://github.com/PowerShell/PowerShell/issues/10983#issuecomment-553759588) this wasn't CDXML related.

The same GetSaferPolicy() is used.

**Comment by eizedev** [comment]
> @JasonFossen I profiled this and verified this is a security check of code we are actually running and if you would like to speed up this check, please adjust the settings as described here:
> [#10983 (comment)](https://github.com/PowerShell/PowerShell/issues/10983#issuecomment-551976417)

For anyone who finds this article here via Google, here's a quick answer on how **we** "solved" it for us, for systems that don't have any internet connection.

Note from me: 
>I spent a few hours looking for a solution to this "problem". As @JasonFossen also mentioned, the startup performance of Powershell 7 discourages new users who want to make a switch from Powershell 5 to Powershell 7 or even coming from Linux/Unix/macOS. (Already experienced myself several times). With a few exceptions (WSUS/SCCM etc.), all of our servers have no Internet connection at all, and many companies will feel the same way.

Another note:
> Could you try this right after you've started with `pwsh -nop -noni`:
> 
> ```
> measure-command { ipmo Microsoft.PowerShell.Management }
> ```
> 
> How long does that take? Then restart with `pwsh -nop -noni` and run:
> 
> ```
> measure-command { ipmo PSReadLine }
> ```

- Result: **15** seconds waiting on importing PSReadLine.


The following two registry keys will reduce the time that Windows waits for a Certificate Revocation List retrieval. This is relevant for connections that require certificate validation (for example, when connecting to a target using SSL). This will also speed up the start time of powershell after setting these two registry keys.

| :zap:        Only enable these settings if your machine isn't exposed to the Internet or doesn't have access to a Certificate Revocation List! Please also keep in mind that this is not a global solution. The problem with CRL checking is different from application to application, from server to server! Please also always think about how to provide your servers with the latest root certificates as well as an up-to-date Certificate Revocation List. |
|-----------------------------------------|

The following two registry keys will reduce the timeout as mentioned above to **1 second**. 

```pwsh
New-ItemProperty -Path "HKLM:\SOFTWARE\Policies\Microsoft\SystemCertificates\ChainEngine\Config" -Name ChainUrlRetrievalTimeoutMilliseconds -Value 1000 -PropertyType DWORD -Force
New-ItemProperty -Path "HKLM:\SOFTWARE\Policies\Microsoft\SystemCertificates\ChainEngine\Config" -Name ChainRevAccumulativeUrlRetrievalTimeoutMilliseconds -Value 1000 -PropertyType DWORD -Force
```

You can also specify these settings via `Domain Group Policy` or locally via `gpedit.msc`:

`Local Computer Policy\Computer Configuration\Windows Settings\Security Settings\Public Key Policies`

![image](https://user-images.githubusercontent.com/6794362/108064736-62652f80-705d-11eb-95b1-410358726940.png)

**Comment by eizedev** [comment]
> @JasonFossen I profiled this and verified this is a security check of code we are actually running and if you would like to speed up this check, please adjust the settings as described here:
> [#10983 (comment)](https://github.com/PowerShell/PowerShell/issues/10983#issuecomment-551976417)

For anyone who finds this article here via Google, here's a quick answer on how **we** "solved" it for us, for systems that don't have any internet connection.

Note from me: 
>I spent a few hours looking for a solution to this "problem". As @JasonFossen also mentioned, the startup performance of Powershell 7 discourages new users who want to make a switch from Powershell 5 to Powershell 7 or even coming from Linux/Unix/macOS. (Already experienced myself several times). With a few exceptions (WSUS/SCCM etc.), all of our servers have no Internet connection at all, and many companies will feel the same way.

Another note:
> Could you try this right after you've started with `pwsh -nop -noni`:
> 
> ```
> measure-command { ipmo Microsoft.PowerShell.Management }
> ```
> 
> How long does that take? Then restart with `pwsh -nop -noni` and run:
> 
> ```
> measure-command { ipmo PSReadLine }
> ```

- Result: **15** seconds waiting on importing PSReadLine.


The following two registry values will reduce the time that Windows waits for a Certificate Revocation List retrieval. This is relevant for connections that require certificate validation (for example, when connecting to a target using SSL). This will also speed up the start time of powershell after setting these two registry keys.

| :zap:        Only enable these settings if your machine isn't exposed to the Internet or doesn't have access to a Certificate Revocation List! Please also keep in mind that this is not a global solution. The problem with CRL checking is different from application to application, from server to server! Please also always think about how to provide your servers with the latest root certificates as well as an up-to-date Certificate Revocation List. |
|-----------------------------------------|

The following registry keys and values will reduce the timeout as mentioned above to **1 second**. 

```pwsh
# Create the keys if missing 
If((Test-Path 'HKLM:\SOFTWARE\Policies\Microsoft\SystemCertificates\ChainEngine') -eq $false ) { New-Item -Path 'HKLM:\SOFTWARE\Policies\Microsoft\SystemCertificates\ChainEngine' -Force -ErrorAction SilentlyContinue }
If((Test-Path 'HKLM:\SOFTWARE\Policies\Microsoft\SystemCertificates\ChainEngine\Config') -eq $false ) { New-Item -Path 'HKLM:\SOFTWARE\Policies\Microsoft\SystemCertificates\ChainEngine\Config' -Force -ErrorAction SilentlyContinue }

# Set Timeout values to 1 second (1000 ms)
New-ItemProperty -Path "HKLM:\SOFTWARE\Policies\Microsoft\SystemCertificates\ChainEngine\Config" -Name ChainUrlRetrievalTimeoutMilliseconds -Value 1000 -PropertyType DWORD -Force
New-ItemProperty -Path "HKLM:\SOFTWARE\Policies\Microsoft\SystemCertificates\ChainEngine\Config" -Name ChainRevAccumulativeUrlRetrievalTimeoutMilliseconds -Value 1000 -PropertyType DWORD -Force
```

You can also specify these settings via `Domain Group Policy` or locally via `gpedit.msc`:

`Local Computer Policy\Computer Configuration\Windows Settings\Security Settings\Public Key Policies`

![image](https://user-images.githubusercontent.com/6794362/108064736-62652f80-705d-11eb-95b1-410358726940.png)

**Comment by TravisEz13** [comment]
This is a nice write up of how to get perf for disconnected systems for powershell.  Perhaps we should have something like this in the docs.

**Comment by iSazonov** [comment]
@theJasonHelmick @sdwheeler It seems we still haven't the docs.

**Comment by sdwheeler** [comment]
@TravisEz13 Do we really want to document how CRL checking?

**Comment by iSazonov** [comment]
It is related not only CRL. We have some issues where we say about telemetry disable, automounting, root directory enumeration, AV/AMSI... maybe something else...

**Comment by xtqqczze** [comment]
@microsoft-github-policy-service bad bot

**Comment by sdwheeler** [comment]
Adding documentation in https://github.com/MicrosoftDocs/PowerShell-Docs/pull/12557

**Comment by kilasuit** [comment]
As per this comment I am closing this as Resoultion-Answered

> 

> Adding documentation in [MicrosoftDocs/PowerShell-Docs#12557](https://github.com/MicrosoftDocs/PowerShell-Docs/pull/12557)
